### PR TITLE
[offscreen-canvas] Correctly report result of asynchronous operations

### LINKED ISSUE #20180: [offscreen-canvas] Procedurally-generated tests are invalid
Many of the tests which are generated programmatically report results before any assertions are made. The tests' "template" code defines a single subtest:

    var t = async_test("%(escaped_desc)s");
    t.step(function() {

    var offscreenCanvas = new OffscreenCanvas(100, 50);
    var ctx = offscreenCanvas.getContext('2d');

    %(code)s
    t.done();

    });

Although declared using the `async_test` API, the invocation of the `t.done` method in the test body makes these tests fully synchronous. However, many of the code excerpts used to expand this template include asynchronous operations, e.g. [`compositing/2d.composite.canvas.copy.html`](https://github.com/web-platform-tests/wpt/blob/192a7c9ae8683984ef180d1bd9edc57b2f000072/offscreen-canvas/compositing/2d.composite.canvas.copy.html)

In the absence of additional subtests, asynchronous code cannot influence the results reported by testharness.js.

There are many ways this could be addressed, but the most straightforward may be to relocate the invocation of `t.done` from the template and into each of the code excerpts. This would also be a good opportunity to wrap asynchronously-executed functions with the `t.step_func` method (otherwise, test failures will be reported as harness errors).

/cc @AmeliaBR @annevk @kenrussell @fserb

### PR DESCRIPTION
**Pull request by jugglinmike**
This resolves gh-20180.

In `master` today, the test results are misrepresenting implementation status. Many failing tests are incorrectly reporting as having passed. Some of them also report an error in the testing harness (which ought to hint that something is going wrong), but many claim to be `OK`.

This patch makes the tests accurately report `PASS` or `FAIL` and removes the test harness errors.

To demonstrate this, I scheduled trials for the patch in Chrome and Firefox. We can use https://wpt.fyi to compare the results reported for each browser with and without this patch (hint: there is a toggle button in the UI just below the icon labeled "diff" that helps you to view only differences):

- Firefox https://wpt.fyi/results/offscreen-canvas?diff&filter=ADC&run_id=358480001&run_id=343560004
- Chrome https://wpt.fyi/results/offscreen-canvas?diff&filter=ADC&run_id=362270001&run_id=358460003

Firefox does not implement OffscreenCanvas, so (as expected) this patch had no effect on Firefox's results. There is one discrepancy, but that's due to an unrelated crash--something that happens from time to time in the test environment.

Chrome's results are much improved. Individual test results changed in one of two ways, as reported in the wpt.fyi UI:

- `+1 / -1 / 0` - this means there is one more passing "statistic" and one more failing "statistic." The statistics combine subtest results and harness status, so this specifically means: the harness status changed from `ERROR` to `OK` and the subtest result changed from `PASS` to `FAIL`.
- `0 / -1 / 0` - here, the harness status was unchanged, and the subtest result changed from `PASS` to `FAIL`.

I've split this into two commits: one for the template and source file changes, and one for the corresponding changes to the generated test material. I don't know how you folks prefer to review or merge changes like this, though, and I'm happy to reconfigure as you please.

### CODE REVIEW

File: offscreen-canvas/tools/tests2d.yaml [hunk 1]
```diff
@@ -9663,8 +10139,8 @@
     - CanvasTest
   code: |
     ctx.font = '50px CanvasTest';
-    deferTest();
-    step_timeout(t.step_func_done(function () {
+    new Promise(function(resolve) { step_timeout(resolve, 500); })
+      .then(function() {
```
Comment (jdm) [inline offscreen-canvas/tools/tests2d.yaml]: Shouldn't promise functions like this use t.step_func so the assertions execute within the test harness wrapper?

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by jdm** [comment]
I can't help wondering if we should have different templates for synchronous vs. asynchronous tests, just to avoid so much duplication.

File: offscreen-canvas/tools/tests2d.yaml (hunk 1 above)
Comment (jugglinmike) [reply offscreen-canvas/tools/tests2d.yaml]: We could wrap the handlers, but as long as we're willing to trust the Promise implementation, we'll detect and report the failure with the final invocation on the Promise chain: `.then(t_pass, t_fail)`

**Comment by jugglinmike** [comment]
> I can't help wondering if we should have different templates for synchronous vs. asynchronous tests, just to avoid so much duplication.

I wondered that, too. I decided against it because I saw that [the test files are hard-coded](https://github.com/web-platform-tests/wpt/blob/56450ba3cf3b14ced81df2c268eb9e515d01c5c5/2dcontext/tools/gentestutils.py#L140), but now I can see that [there's a specific provision for this test suite](https://github.com/web-platform-tests/wpt/blob/56450ba3cf3b14ced81df2c268eb9e515d01c5c5/2dcontext/tools/gentestutils.py#L141-L142). I'll see what I can do about this...

File: offscreen-canvas/tools/tests2d.yaml (hunk 1 above)
Comment (jdm) [reply offscreen-canvas/tools/tests2d.yaml]: Ah, that's a good point!

**Comment by jugglinmike** [comment]
@jdm It looks like using multiple templates will require more effort than extending the list in that script. I won't have the bandwidth to do that any time soon, but I would like to see these tests made technically correct. Would you be willing to tolerate the duplication as technical debt?

**Comment by jdm** [comment]
Yes, I won't insist on that change when it's holding back actual test correctness issues.

**Review by jdm — APPROVED** [review]

**Comment by guest271314** [comment]
Note, https://github.com/web-platform-tests/wpt/pull/20277/commits/dce0fc92eaec530de5aa7bbdbf59f5e7acdc9194#diff-8bf796e810f7184850d920e92c11bad2R46 currently will throw an exception at Firefox and Nightly whether the test is synchronous or asynchronous (https://github.com/web-platform-tests/wpt/issues/19799).

**Comment by guest271314** [comment]
Note also, the messages printed at https://wpt.fyi/results/offscreen-canvas/the-offscreen-canvas/2d.getcontext.exists.html?diff&filter=ADC&run_id=358480001&run_id=343560004 for Firefox



> The 2D context is implemented | NOT_DEFINED(OffscreenCanvas) | NOT_DEFINED(OffscreenCanvas) 

and 


> The 2D context is implemented | FAIL message: OffscreenCanvas is not defined | FAIL message: OffscreenCanvas is not defined



are incorrect. `OffscreenCanvas` is defined, `OffscreenCanvasRenderingContext2D` are not defined at Firefox 69 and Nightly 72.

**Comment by guest271314** [comment]
Is the Firefox version being tested launched with `gfx.offscreencanvas.enabled` preference set to `true`?

**Comment by guest271314** [comment]
> Firefox does not implement OffscreenCanvas

Currently Firefox requires the preference `gfx.offscreen.enabled` to be set to true for `OffscreenCanvas` to be defined in main thread and `Worker` scope. Firefox tracking bug https://bugzilla.mozilla.org/show_bug.cgi?id=1390089.

If no preferences were passed to Firefox being used the above quote would be correct and the test re Firefox moot, however, the browser used by WPT does pass custom preferences to the browser instance https://github.com/web-platform-tests/wpt.fyi/issues/1648#issuecomment-558776559. Therefore `gfx.offscreen.enabled` can be set to `true` to actually test the current implementation of `OffscreenCanvas` at Firefox.

**Comment by jugglinmike** [comment]
@guest271314 Thanks, that's all good information. This pull request is focused on correcting the tests themselves, so discussion about configuring the browsers is probably better left to a separate issue.

**Comment by jugglinmike** [comment]
The stability check in Firefox appears to be timing out due to the number of tests modified by this patch. @jdm if you could verify my interpretation, then I will override the warning about the failed job.

**Comment by jdm** [comment]
Agreed!

**Pull request merged**